torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -768,6 +768,15 @@ def get_state_dict_type(module: nn.Module) -> StateDictSettings:
                     state_dict_config=submodule._state_dict_config,
                     optim_state_dict_config=submodule._optim_state_dict_config,
                 )
+
+                # If device_mesh is passed in when initalizing FSDP, we automatically turn the
+                # _use_dtensor flag to be true for ShardedOptimStateDictConfig().
+                if (
+                    getattr(submodule, "device_mesh", None)
+                    and state_dict_settings.state_dict_type
+                    == StateDictType.SHARDED_STATE_DICT
```

Comment:
If this is a requirement, should `state_dict_type` be asserted rather than checked?
E.g.
```
if getattr(submodule, "device_mesh", None):
  assert state_dict_settings.state_dict_type == StateDictType.SHARDED_STATE_DICT
  state_dict_settings.optim_state_dict_config._use_dtensor = True
```

Reply:
> If this is a requirement, should `state_dict_type` be asserted rather than checked? E.g.
> 
> ```
> if getattr(submodule, "device_mesh", None):
>   assert state_dict_settings.state_dict_type == StateDictType.SHARDED_STATE_DICT
>   state_dict_settings.optim_state_dict_config._use_dtensor = True
> ```

The reason we didn't assert here is that users can still choose to use full_state_dict. Since other than getting the DTensor state_dict, DeviceMesh can make HSDP initialization much more easily.

This is a good point though. Let me discuss it with the team to see if we want to only allow sharded_state_dict when using DeviceMesh.

Reply:
I see. Maybe this can explicitly check if type is `FULL_STATE_DICT` and if so dump some logging info saying dtensor is not used because of this (as opposed to silently diverting from the regular device mesh -> dtensor enabled logic).
Otherwise assert `SHARDED` must be used instead.